[Input] Improve Input System action migration from InputManager.asset file

AlexTyrer opened the pull request:
### Description
Improved Input System action migration from InputManager.asset file

### Changes made

o Ensure InputSystem.actions is set after migrating InputActionAssets
o Ensure AssetDatabase is aware of newly created .inputaction asset
o Don't overwrite existing Input System action file - always migrate settings to a new unique file

### Notes

_Please write down any additional notes, remove the section if not applicable._

### Checklist

Before review:

- Changelog entry added
- No tests added - required manual testing with different Input System (pre-release) versions
- No documentation required

During merge:

(notes for later...)
CHANGE: Migrate any project-wide input actions found in the InputManager.asset file to a new InputSystem_Actions.inputactions asset file.

AlexTyrer committed:
[Input] Improve Input System action migration from InputManager.asset file

o Ensure InputSystem.actions is set after migrating migratedInputActionAssets
o Ensure AssetDatabase is aware of newly created .inputaction asset
o Don't overwrite existing Input System action file - always migrate settings to a new unique file

AlexTyrer committed:
[Input] Format source code (ProjectWideActions.cs)

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -39,14 +34,39 @@ static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAsse
+                            path = "Assets/" + unique + ".inputactions";
+                        }
+
+                        var json = inputActionsAsset.ToJson();
```
This line fixed the JSON data and was critical :)

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -39,14 +34,39 @@ static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAsse
+                        File.WriteAllText(path, json);
+
+                        AssetDatabase.Refresh();
+                        InputSystem.actions = (InputActionAsset)AssetDatabase.LoadAssetAtPath(path, typeof(InputActionAsset));
```
Excellent - this makes it show up now in project settings

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -63,6 +83,8 @@ internal static void MoveInputManagerAssetActionsToProjectWideInputActionAsset()
                         }
                     }
+
+                    AssetDatabase.SaveAssets();
```
Great - this made the InputManager.asset file change at this point so we could see it cleaned up

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -39,14 +34,39 @@ static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAsse
+                                names.Add(System.IO.Path.GetFileNameWithoutExtension(files[i]));
+                            }
+                            string unique = ObjectNames.GetUniqueName(names.ToArray(), "InputSystem_Actions");
+                            path = "Assets/" + unique + ".inputactions";
```
Great - this prevents overwriting user files

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -15,15 +16,9 @@ internal static class ProjectWideActionsAsset
         internal static class ProjectSettingsProjectWideActionsAssetConverter
         {
-            internal const string kAssetPath = "ProjectSettings/InputManager.asset";
+            internal const string kAssetPathInputManager = "ProjectSettings/InputManager.asset";
```
Easier to read this name below - thanks

ekcoh replied:
+1 better name for what it implies

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -15,15 +16,9 @@ internal static class ProjectWideActionsAsset
         {
-            internal const string kAssetPath = "ProjectSettings/InputManager.asset";
+            internal const string kAssetPathInputManager = "ProjectSettings/InputManager.asset";
             internal const string kAssetName = InputSystem.kProjectWideActionsAssetName;
```
Could name this kAssetNameProjectWide but not critical

ekcoh replied:
Not sure why defining a constant set to another constant? However...
InputSystem.kProjectWideActionsAssetName should no longer exist, so rather move that constant here instead. Its only relevant for the previous solution. It seems to not be used anywhere else....

AlexTyrer replied:
That's probably a result of starting in trunk and then moving to your branch - will define the constant here and remove the old one in InputSystem.cs

ekcoh replied:
Sounds good.

lyndon-unity reviewed (approved):
Great update - much more robust

AlexTyrer committed:
Merge branch 'develop' into ISX-1867-project-wide-asset-migration

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -39,14 +34,39 @@ static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAsse
             internal static void MoveInputManagerAssetActionsToProjectWideInputActionAsset()
             {
-                var objects = AssetDatabase.LoadAllAssetsAtPath(kAssetPath);
+                var objects = AssetDatabase.LoadAllAssetsAtPath(kAssetPathInputManager);
```
I think this needs to use FileUtils.PhysicalPath to be future proof, DM me for further elaboration. See EditorHelpers.cs for how its used conditionally. I have extracted the #ifdef into its own function on a pending PR but its not available in this version.

AlexTyrer replied:
Just change to this?

```
#if UNITY_2021_2_OR_NEWER
                var objects = AssetDatabase.LoadAllAssetsAtPath(FileUtil.GetPhysicalPath(kAssetPathInputManager));
#else
                var objects = AssetDatabase.LoadAllAssetsAtPath(kAssetPathInputManager);
#endif
```

AlexTyrer replied:
Will make this change for the other file path too - will a local conversion function to avoid so many pre-processor blocks.

```
                        string path = ProjectWideActionsAsset.kDefaultAssetPath;

                        if (File.Exists(path))

```

ekcoh replied:
Sounds good

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -39,14 +34,39 @@ static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAsse
+                        //
+                        string path = ProjectWideActionsAsset.kDefaultAssetPath;
+
+                        if (File.Exists(path))
```
Minor: Would suggest doing inverted ifs in this function in general if all of them are early-outs since there is a lot of nesting making it harder to read.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -39,14 +34,39 @@ static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAsse
+                                names.Add(System.IO.Path.GetFileNameWithoutExtension(files[i]));
+                            }
+                            string unique = ObjectNames.GetUniqueName(names.ToArray(), "InputSystem_Actions");
+                            path = "Assets/" + unique + ".inputactions";
```
Great to see the "finding an available name" logic to avoid the clash. Would however recommend using InputActionImporter.cs constant for extension though rather than hardcoding a string copy here. That way this just works if it would change in the future.

AlexTyrer replied:
I can add a constant, but there's already some duplication here:

        private const string kDefaultAssetPath = "Assets/InputSystem_Actions.inputactions";

Do you mean to define something like:

        private const string kInputActionsBaseName = InputSystem_Action";

AlexTyrer replied:
I've tidied these names up.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -39,14 +34,39 @@ static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAsse
+                        }
+
+                        var json = inputActionsAsset.ToJson();
+                        File.WriteAllText(path, json);
```
I would suggest removing this File.WriteAllText + AssetDatabase.Refresh() with a call to InputActionAssetManager.SaveAsset instead. In addition that method handles any underlying single-checkout versioning system which could otherwise generate a File-access exception here.

AlexTyrer replied:
Will make this change..

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs:
```diff
@@ -39,14 +34,39 @@ static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAsse
+                        File.WriteAllText(path, json);
+
+                        AssetDatabase.Refresh();
+                        InputSystem.actions = (InputActionAsset)AssetDatabase.LoadAssetAtPath(path, typeof(InputActionAsset));
```
I wonder if we shouldn't provide a Debug.Info message here informing about the conversion that have been done and that the asset has been assigned to InputSystem.actions?

Also, what if InputSystem.actions is already assigned? In that case I do not think we should assign it, only port to a new asset and inform about the conversion that has been made.

AlexTyrer replied:
Happy to add a message, but I can't find any example of "Debug.Info" - can you provide an example?

As for the assignment, we are in this block of code because we found project wide settings in the input manager asset file.

When testing this with @lyndon-unity we worked on the assumption that the user probably did want the assignment to be updated.

ekcoh replied:
Thinking about this, maybe just let it be as is

AlexTyrer replied:
I've added a Debug.Log() for both of these: migration (and assignment if required)

ekcoh reviewed (changes requested):
Good improvement of the import/porting logic. I noticed however that the AssetProcessor in this file is using the callback with the `didDomainReload` argument. This is not present in versions less than Unity 2021.2. I did a similar fix recently to support compilation for 2019, see https://github.com/Unity-Technologies/InputSystem/pull/1848
I advice to do a similar change here, and I cannot understand how this could pass CI without this being updated on that PR, do you know? Or was the fix I did unnecessary somehow and was a false positive form IDE? (Confused)

ekcoh reviewed (commented):
This branch is missing a CHANGELOG.md entry so that needs to be added.

I also wonder whether it would be feasible to add a unit test exercising the porting logic? It would be helpful to avoid any future problems. I believe that could be done by writing an asset containing a InputManager.asset copy to any test path and verify the conversion result and that it is cleaned up. Such a test would of course be editor only so conditionally compiled for UNITY_EDITOR.

AlexTyrer commented:
> Good improvement of the import/porting logic. I noticed however that the AssetProcessor in this file is using the callback with the `didDomainReload` argument. This is not present in versions less than Unity 2021.2. I did a similar fix recently to support compilation for 2019, see #1848 I advice to do a similar change here, and I cannot understand how this could pass CI without this being updated on that PR, do you know? Or was the fix I did unnecessary somehow and was a false positive form IDE? (Confused)

I am no sure either.

I see other instances of functions with this parameter in the repo (without a  #if UNITY_2021_2_OR_NEWER conditional)
eg
**RenderPipelineGlobalSettingsPostprocessor.cs**

https://github.cds.internal.unity3d.com/unity/unity/blob/15f36247668085dfe54fbf0301812c4e89fff643/Editor/Mono/Graphics/RenderPipelineGlobalSettingsPostprocessor.cs#L75

**TerrainTransientToolbarOverlay.cs**
https://github.cds.internal.unity3d.com/unity/unity/blob/15f36247668085dfe54fbf0301812c4e89fff643/Modules/TerrainEditor/Overlays/TerrainTransientToolbarOverlay.cs#L21

Update: of course these are in the Unity repo - the package needs to be compatible with multiple releases - have added the pre-processor check.

ekcoh commented:
Might have run into instabilities with CI, recommend opening All Tests in Yamato and reschedule failed jobs.

ekcoh commented:
Approved under condition that CHANGELOG.md is added and CI is green.

AlexTyrer committed:
[Input] PR feedback - https://github.com/Unity-Technologies/InputSystem/pull/1849

o Cleaned-up naming constants (removed obsolete InputSystem.kProjectWideActionsAssetName)
o Added UNITY_2021_2_OR_NEWER checks / alternative code-paths for path related API changes: added FixPath() function
o Replaced File.WriteAllText()/AssetDatabase.Refresh() with InputActionAssetManager.SaveAsset()
o Added some Debug.Log() showing asset migration / Project-wide Input Actions assignment
o Reduced code indentation with an early out
o Updated changelog

AlexTyrer committed:
Merge branch 'ISX-1867-project-wide-asset-migration' of https://github.com/Unity-Technologies/InputSystem into ISX-1867-project-wide-asset-migration

# Conflicts:
#	Packages/com.unity.inputsystem/CHANGELOG.md

AlexTyrer committed:
Merge branch 'develop' into ISX-1867-project-wide-asset-migration

Pauliusd01 commented:
Checking

Pauliusd01 commented:
Two things I noticed when upgrading an older 1.8.0-pre.2 project:

- It correctly creates a mirrored projectwideactions asset for you but that asset reference is not updated in any of the playerinput components, that component still points to the old, now non-existent projectwideactions. Should those automatically update? (See video, the name of the asset referenced is that of the old one I have to manually select the one your PR created for me)

https://github.com/Unity-Technologies/InputSystem/assets/54306142/aea93302-7e03-4ccd-a167-7155d261064d

- The newly created asset is dirty on creation, I was prompted to save "changes" right after migrating and opening it

I don't think either is a blocker but just mentioning any odd things I come across

Pauliusd01 reviewed (approved):
LGTM, did not notice any major issues with upgrading projects that use ProjectWideActions or existing functionality. Only left 2 minor notes above for consideration

AlexTyrer committed:
Merge branch 'develop' into ISX-1867-project-wide-asset-migration

AlexTyrer committed:
Merge branch 'develop' into ISX-1867-project-wide-asset-migration
